File: 02-unidade/C++/ArvoreRubroNegra/ArvoreRubroNegra.cpp
#include <iostream>
#include <queue>

using namespace std;

struct NoRB {
    int item;
    string cor;
    NoRB *esquerda, *direita, *pai;

    NoRB(int chave) : item(chave), cor("Vermelho"), esquerda(nullptr), direita(nullptr), pai(nullptr) {}
};

class ArvoreRubroNegra {
private:
    NoRB* TNULL;
    NoRB* root;

    void inicializarTNULL() {
        TNULL = new NoRB(0);
        TNULL->cor = "Preto";
        TNULL->esquerda = nullptr;
        TNULL->direita = nullptr;
    }

    void rotacaoEsquerda(NoRB* x) {
        NoRB* y = x->direita;
        x->direita = y->esquerda;
        if (y->esquerda != TNULL) {
            y->esquerda->pai = x;
        }
        y->pai = x->pai;
        if (x->pai == nullptr) {
            root = y;
        } else if (x == x->pai->esquerda) {
            x->pai->esquerda = y;
        } else {
            x->pai->direita = y;
        }
        y->esquerda = x;
        x->pai = y;
    }

    void rotacaoDireita(NoRB* x) {
        NoRB* y = x->esquerda;
        x->esquerda = y->direita;
        if (y->direita != TNULL) {
            y->direita->pai = x;
        }
        y->pai = x->pai;
        if (x->pai == nullptr) {
            root = y;
        } else if (x == x->pai->direita) {
            x->pai->direita = y;
        } else {
            x->pai->esquerda = y;
        }
        y->direita = x;
        x->pai = y;
    }

    void corrigirInsercao(NoRB* k) {
        while (k->pai && k->pai->cor == "Vermelho") {
            if (k->pai == k->pai->pai->esquerda) {
                NoRB* tio = k->pai->pai->direita;
                if (tio && tio->cor == "Vermelho") {
                    k->pai->cor = "Preto";
                    tio->cor = "Preto";
                    k->pai->pai->cor = "Vermelho";
                    k = k->pai->pai;
                } else {
                    if (k == k->pai->direita) {
                        k = k->pai;
                        rotacaoEsquerda(k);
                    }
                    k->pai->cor = "Preto";
                    k->pai->pai->cor = "Vermelho";
                    rotacaoDireita(k->pai->pai);
                }
            } else {
                NoRB* tio = k->pai->pai->esquerda;
                if (tio && tio->cor == "Vermelho") {
                    k->pai->cor = "Preto";
                    tio->cor = "Preto";
                    k->pai->pai->cor = "Vermelho";
                    k = k->pai->pai;
                } else {
                    if (k == k->pai->esquerda) {
                        k = k->pai;
                        rotacaoDireita(k);
                    }
                    k->pai->cor = "Preto";
                    k->pai->pai->cor = "Vermelho";
                    rotacaoEsquerda(k->pai->pai);
                }
            }
        }
        root->cor = "Preto";
    }

    void corrigirRemocao(NoRB* x) {
        while (x != root && x->cor == "Preto") {
            if (x == x->pai->esquerda) {
                NoRB* s = x->pai->direita;
                if (s->cor == "Vermelho") {
                    s->cor = "Preto";
                    x->pai->cor = "Vermelho";
                    rotacaoEsquerda(x->pai);
                    s = x->pai->direita;
                }
                if (s->esquerda->cor == "Preto" && s->direita->cor == "Preto") {
                    s->cor = "Vermelho";
                    x = x->pai;
                } else {
                    if (s->direita->cor == "Preto") {
                        s->esquerda->cor = "Preto";
                        s->cor = "Vermelho";
                        rotacaoDireita(s);
                        s = x->pai->direita;
                    }
                    s->cor = x->pai->cor;
                    x->pai->cor = "Preto";
                    s->direita->cor = "Preto";
                    rotacaoEsquerda(x->pai);
                    x = root;
                }
            } else {
                NoRB* s = x->pai->esquerda;
                if (s->cor == "Vermelho") {
                    s->cor = "Preto";
                    x->pai->cor = "Vermelho";
                    rotacaoDireita(x->pai);
                    s = x->pai->esquerda;
                }
                if (s->direita->cor == "Preto" && s->esquerda->cor == "Preto") {
                    s->cor = "Vermelho";
                    x = x->pai;
                } else {
                    if (s->esquerda->cor == "Preto") {
                        s->direita->cor = "Preto";
                        s->cor = "Vermelho";
                        rotacaoEsquerda(s);
                        s = x->pai->esquerda;
                    }
                    s->cor = x->pai->cor;
                    x->pai->cor = "Preto";
                    s->esquerda->cor = "Preto";
                    rotacaoDireita(x->pai);
                    x = root;
                }
            }
        }
        x->cor = "Preto";
    }

    void substituirNo(NoRB* u, NoRB* v) {
        if (u->pai == nullptr) {
            root = v;
        } else if (u == u->pai->esquerda) {
            u->pai->esquerda = v;
        } else {
            u->pai->direita = v;
        }
        v->pai = u->pai;
    }

    void inOrder(NoRB* no) {
        if (no != TNULL) {
            inOrder(no->esquerda);
            cout << no->item << " (" << no->cor << ") ";
            inOrder(no->direita);
        }
    }

public:
    ArvoreRubroNegra() {
        inicializarTNULL();
        root = TNULL;
    }

    void inserir(int chave) {
        NoRB* novoNo = new NoRB(chave);
        novoNo->esquerda = TNULL;
        novoNo->direita = TNULL;

        NoRB* y = nullptr;
        NoRB* x = root;

        while (x != TNULL) {
            y = x;
            if (novoNo->item < x->item) {
                x = x->esquerda;
            } else {
                x = x->direita;
            }
        }

        novoNo->pai = y;
        if (y == nullptr) {
            root = novoNo;
        } else if (novoNo->item < y->item) {
            y->esquerda = novoNo;
        } else {
            y->direita = novoNo;
        }

        if (novoNo->pai == nullptr) {
            novoNo->cor = "Preto";
            return;
        }

        if (novoNo->pai->pai == nullptr) {
            return;
        }

        corrigirInsercao(novoNo);
    }

    void remover(int chave) {
        NoRB* z = buscar(chave);
        if (z == TNULL) return;

        NoRB* y = z;
        NoRB* x;
        string yCorOriginal = y->cor;

        if (z->esquerda == TNULL) {
            x = z->direita;
            substituirNo(z, z->direita);
        } else if (z->direita == TNULL) {
            x = z->esquerda;
            substituirNo(z, z->esquerda);
        } else {
            y = minimo(z->direita);
            yCorOriginal = y->cor;
            x = y->direita;
            if (y->pai == z) {
                x->pai = y;
            } else {
                substituirNo(y, y->direita);
                y->direita = z->direita;
                y->direita->pai = y;
            }
            substituirNo(z, y);
            y->esquerda = z->esquerda;
            y->esquerda->pai = y;
            y->cor = z->cor;
        }

        if (yCorOriginal == "Preto") {
            corrigirRemocao(x);
        }
    }

    NoRB* buscar(int chave) {
        NoRB* atual = root;
        while (atual != TNULL && chave != atual->item) {
            if (chave < atual->item) {
                atual = atual->esquerda;
            } else {
                atual = atual->direita;
            }
        }
        return atual;
    }

    NoRB* minimo(NoRB* no) {
        while (no->esquerda != TNULL) {
            no = no->esquerda;
        }
        return no;
    }

    void emOrdem() {
        if (root == TNULL) {
            cout << "Arvore está vazia." << endl;
            return;
        }
        inOrder(root);
        cout << endl;
    }
};

int main() {
    ArvoreRubroNegra arvore;
    int opcao = 0;

    arvore.inserir(15);
    arvore.inserir(18);
    arvore.inserir(20);
    arvore.inserir(35);
    arvore.inserir(32);
    arvore.inserir(38);
    arvore.inserir(30);
    arvore.inserir(40);
    arvore.inserir(32);
    arvore.inserir(45);
    arvore.inserir(48);
    arvore.inserir(52);
    arvore.inserir(60);
    arvore.inserir(50);

    while (opcao != 5) {
        cout << "*************************************" << endl;
        cout << "Entre com a opcao:" << endl;
        cout << "** 1 **: Inserir valor" << endl;
        cout << "** 2 **: Remover valor" << endl;
        cout << "** 3 **: Buscar valor" << endl;
        cout << "** 4 **: Exibir Arvore" << endl;
        cout << "** 5 **: Sair do programa" << endl;
        cout << "*************************************" << endl;
        cout << "-> ";
        cin >> opcao;

        if (opcao == 1) {
            int x;
            cout << "Informe o valor -> ";
            cin >> x;
            arvore.inserir(x);
        } else if (opcao == 2) {
            int x;
            cout << "Informe o valor -> ";
            cin >> x;
            arvore.remover(x);
        } else if (opcao == 3) {
            int x;
            cout << "Informe o valor -> ";
            cin >> x;
            NoRB* resultado = arvore.buscar(x);
            if (resultado != nullptr) {
                cout << "Valor " << x << " encontrado na arvore." << endl;
            } else {
                cout << "Valor " << x << " nao encontrado na arvore." << endl;
            }
        } else if (opcao == 4) {
            arvore.emOrdem();
        } else if (opcao == 5) {
            cout << "Saindo do programa..." << endl;
            break;
        } else {
            cout << "Opcao invalida! Tente novamente." << endl;
        }
    }

    return 0;
}

File: 02-unidade/Python/Arvore Rubro Negra/ArvoreRubroNegra.py
from ArvoreRubroNegroNo import NoRB


class ArvoreRubroNegra:
     def __init__(self):
        self.TNULL = NoRB(None)
        self.TNULL.cor = "Preto"
        self.root = self.TNULL

     def rotacao_esquerda(self, no):
        y = no.direita
        no.direita = y.esquerda
        if y.esquerda != self.TNULL:
            y.esquerda.pai = no

        y.pai = no.pai
        if no.pai is None:
            self.root = y
        elif no == no.pai.esquerda:
            no.pai.esquerda = y
        else:
            no.pai.direita = y

        y.esquerda = no
        no.pai = y

     def rotacao_direita(self, no):
        y = no.esquerda
        no.esquerda = y.direita
        if y.direita != self.TNULL:
            y.direita.pai = no

        y.pai = no.pai
        if no.pai is None:
            self.root = y
        elif no == no.pai.direita:
            no.pai.direita = y
        else:
            no.pai.esquerda = y

        y.direita = no
        no.pai = y

     def corrigir_insercao(self, k):
        while k.pai and k.pai.cor == "Vermelho":
            if k.pai == k.pai.pai.esquerda:
                tio = k.pai.pai.direita
                if tio and tio.cor == "Vermelho":
                    k.pai.cor = "Preto"
                    tio.cor = "Preto"
                    k.pai.pai.cor = "Vermelho"
                    k = k.pai.pai
                else:
                    if k == k.pai.direita:
                        k = k.pai
                        self.rotacao_esquerda(k)
                    k.pai.cor = "Preto"
                    k.pai.pai.cor = "Vermelho"
                    self.rotacao_direita(k.pai.pai)
            else:
                tio = k.pai.pai.esquerda
                if tio and tio.cor == "Vermelho":
                    k.pai.cor = "Preto"
                    tio.cor = "Preto"
                    k.pai.pai.cor = "Vermelho"
                    k = k.pai.pai
                else:
                    if k == k.pai.esquerda:
                        k = k.pai
                        self.rotacao_direita(k)
                    k.pai.cor = "Preto"
                    k.pai.pai.cor = "Vermelho"
                    self.rotacao_esquerda(k.pai.pai)
        self.root.cor = "Preto"

     def corrigir_remocao(self, x):
        while x != self.root and x.cor == "Preto":
            if x == x.pai.esquerda:
                s = x.pai.direita
                if s.cor == "Vermelho":
                    s.cor = "Preto"
                    x.pai.cor = "Vermelho"
                    self.rotacao_esquerda(x.pai)
                    s = x.pai.direita
                if s.esquerda.cor == "Preto" and s.direita.cor == "Preto":
                    s.cor = "Vermelho"
                    x = x.pai
                else:
                    if s.direita.cor == "Preto":
                        s.esquerda.cor = "Preto"
                        s.cor = "Vermelho"
                        self.rotacao_direita(s)
                        s = x.pai.direita
                    s.cor = x.pai.cor
                    x.pai.cor = "Preto"
                    s.direita.cor = "Preto"
                    self.rotacao_esquerda(x.pai)
                    x = self.root
            else:
                s = x.pai.esquerda
                if s.cor == "Vermelho":
                    s.cor = "Preto"
                    x.pai.cor = "Vermelho"
                    self.rotacao_direita(x.pai)
                    s = x.pai.esquerda
                if s.direita.cor == "Preto" and s.esquerda.cor == "Preto":
                    s.cor = "Vermelho"
                    x = x.pai
                else:
                    if s.esquerda.cor == "Preto":
                        s.direita.cor = "Preto"
                        s.cor = "Vermelho"
                        self.rotacao_esquerda(s)
                        s = x.pai.esquerda
                    s.cor = x.pai.cor
                    x.pai.cor = "Preto"
                    s.esquerda.cor = "Preto"
                    self.rotacao_direita(x.pai)
                    x = self.root
        x.cor = "Preto"

     def substituir_no(self, u, v):
        if u.pai is None:
            self.root = v
        elif u == u.pai.esquerda:
            u.pai.esquerda = v
        else:
            u.pai.direita = v
        v.pai = u.pai

     def remover(self, chave):
        no = self.buscar(chave)
        if no == self.TNULL:
            return

        y = no
        y_cor_original = y.cor
        if no.esquerda == self.TNULL:
            x = no.direita
            self.substituir_no(no, no.direita)
        elif no.direita == self.TNULL:
            x = no.esquerda
            self.substituir_no(no, no.esquerda)
        else:
            y = self.minimo(no.direita)
            y_cor_original = y.cor
            x = y.direita
            if y.pai == no:
                x.pai = y
            else:
                self.substituir_no(y, y.direita)
                y.direita = no.direita
                y.direita.pai = y

            self.substituir_no(no, y)
            y.esquerda = no.esquerda
            y.esquerda.pai = y
            y.cor = no.cor

        if y_cor_original == "Preto":
            self.corrigir_remocao(x)

     
     def inserir(self, chave):
        novo_no = NoRB(chave)
        novo_no.esquerda = self.TNULL
        novo_no.direita = self.TNULL

        y = None
        x = self.root

        while x != self.TNULL:
            y = x
            if novo_no.item < x.item:
                x = x.esquerda
            else:
                x = x.direita

        novo_no.pai = y
        if y is None:
            self.root = novo_no
        elif novo_no.item < y.item:
            y.esquerda = novo_no
        else:
            y.direita = novo_no

        if novo_no.pai is None:
            novo_no.cor = "Preto"
            return

        if novo_no.pai.pai is None:
            return

        self.corrigir_insercao(novo_no)

     def buscar(self, chave):
          atual = self.root
          while atual != self.TNULL and chave != atual.item:
               if chave < atual.item:
                    atual = atual.esquerda
               else:
                    atual = atual.direita
          return atual
     
     def minimo(self, no):
          while no.esquerda != self.TNULL:
               no = no.esquerda
          return no
     
     def emNivel(self):
          if self.root == self.TNULL:
               print("Árvore está vazia.")
               return

          fila = [self.root]
          while fila:
               atual = fila.pop(0)
               print(f"{atual.item} ({atual.cor})", end=" ")

               if atual.esquerda != self.TNULL:
                    fila.append(atual.esquerda)

               if atual.direita != self.TNULL:
                    fila.append(atual.direita)

     def printarArvore(self):
          print("Árvore em nível:")
          self.emNivel()
          print()
